### PR TITLE
customize used_by association uniqueness

### PR DESCRIPTION
**Pull request by anzioka**
PR:
- Customize user_apps -> usedby -> participant association uniqueness based on app settings

### CODE REVIEW

File: src/main/java/com/openlattice/chronicle/controllers/v2/ChronicleControllerV2.java [hunk 1]
```diff
@@ -202,6 +206,19 @@ public Integer upload(
...
 
+    @Override
+    @RequestMapping(
+            path = ORGANIZATION_ID_PATH + SETTINGS_PATH,
+            method = RequestMethod.GET,
+            produces = MediaType.APPLICATION_JSON_VALUE
+    )
+    public Map<String, Object> getAppSettings(
+            @PathVariable ( ORGANIZATION_ID ) UUID organizationId,
+            @RequestParam (value = APP_NAME) String appName
```
Comment (UnsungHero97) [inline src/main/java/com/openlattice/chronicle/controllers/v2/ChronicleControllerV2.java:217]:
```suggestion
            @PathVariable( ORGANIZATION_ID ) UUID organizationId,
            @RequestParam( value = APP_NAME ) String appName
```

File: src/main/java/com/openlattice/chronicle/services/entitysets/EntitySetIdsService.java [hunk 2]
```diff
@@ -190,6 +200,18 @@ public void refreshLegacyParticipantEntitySetIds() {
...
     }
 
+    // app settings
+    @Override
+    public Map<String, Object> getOrgAppSettings(AppComponent appComponent, UUID organizationId) {
+        return appSettings.getOrDefault( appComponent, ImmutableMap.of() ).getOrDefault( organizationId, ImmutableMap.of() );
+    }
+
+    @Override
+    public Map<String, Object> getOrgAppSettings (String appName, UUID organizationId) {
```
Comment (UnsungHero97) [inline src/main/java/com/openlattice/chronicle/services/entitysets/EntitySetIdsService.java:210]:
```suggestion
    public Map<String, Object> getOrgAppSettings(String appName, UUID organizationId) {
```

File: src/main/java/com/openlattice/chronicle/services/surveys/SurveysService.java [hunk 3]
```diff
@@ -966,7 +966,7 @@ public void submitTimeUseDiarySurvey(
             Map<UUID, Map<UUID, Map<UUID, Set<Object>>>> entitiesByEntitySet = getEntitiesByESID( entityKeyIdMap,
                     entitiesByEntityKey );
             entitiesByEntitySet.forEach( ( entitySetId, groupedEntities ) -> {
-                dataApi.updateEntitiesInEntitySet( entitySetId, groupedEntities, UpdateType.PartialReplace );
+                dataApi.updateEntitiesInEntitySet( entitySetId, groupedEntities, UpdateType.Merge );
```
Comment (UnsungHero97) [inline src/main/java/com/openlattice/chronicle/services/surveys/SurveysService.java:969]: Why are you making this change? What does it fix?

File: src/main/java/com/openlattice/chronicle/services/upload/AppDataUploadService.java [hunk 4]
```diff
@@ -504,6 +520,12 @@ private void createEntitiesAndAssociations(
                 appName = getFirstValueOrNull( appEntity, TITLE_FQN );
             }
 
+            String recordType = getFirstValueOrNull( appEntity, RECORD_TYPE_FQN );
+            if ( recordType == null || !recordType.equals( FOREGROUND_EVENT_TYPE ) ) {
```
Comment (UnsungHero97) [inline src/main/java/com/openlattice/chronicle/services/upload/AppDataUploadService.java:524]: Should that `!` be there?

File: src/main/java/com/openlattice/chronicle/services/surveys/SurveysService.java (hunk 3 above)
Comment (anzioka) [reply src/main/java/com/openlattice/chronicle/services/surveys/SurveysService.java:969]: Actually this change was unnecessary.

**Review by UnsungHero97 — APPROVED** [review]

**Comment by UnsungHero97** [comment]
Blocked by https://github.com/openlattice/chronicle-api/pull/52

**Review by UnsungHero97 — APPROVED** [review]

File: src/main/java/com/openlattice/chronicle/constants/AppComponent.java [hunk 5]
```diff
@@ -18,4 +18,13 @@
     public String toString() {
         return component;
     }
+
+    public static AppComponent fromString(String name) {
```
Comment (geekbeast) [inline src/main/java/com/openlattice/chronicle/constants/AppComponent.java:22]: We dug our own whole here, would have be better to just use default serialization, but I am guessing these are used elsewhere?

File: src/main/java/com/openlattice/chronicle/services/upload/AppDataUploadService.java [hunk 6]
```diff
@@ -494,7 +505,12 @@ private void createEntitiesAndAssociations(
 
             String appPackageName, appName;
             appPackageName = appName = getFirstValueOrNull( appEntity, FULL_NAME_FQN );
-            String dateLogged = getMidnightDateTime( getFirstValueOrNull( appEntity, DATE_LOGGED_FQN ) );
+            String eventDate = getFirstValueOrNull( appEntity, DATE_LOGGED_FQN );
```
Comment (geekbeast) [inline src/main/java/com/openlattice/chronicle/services/upload/AppDataUploadService.java:508]: Is this the same things Objects.firstNotNull or MoreObjects.firstNotNull?

**Review by geekbeast — COMMENTED** [review]
Just a few clarifications

File: src/main/java/com/openlattice/chronicle/services/upload/AppDataUploadService.java (hunk 6 above)
Comment (anzioka) [reply src/main/java/com/openlattice/chronicle/services/upload/AppDataUploadService.java:508]: Nope. This gets returns the first value from a `SetMultimap<UUID, Object> entity` with key propertyTypeId(fqn)

File: src/main/java/com/openlattice/chronicle/constants/AppComponent.java (hunk 5 above)
Comment (anzioka) [reply src/main/java/com/openlattice/chronicle/constants/AppComponent.java:22]: Correct. 

I added this in order to resolve an api request string parameter to one of the enums.

**Review by geekbeast — APPROVED** [review]
It would be nice to clean up the enum, but understand it is not a priority.

**Pull request merged**